// @wraith13/vscel/library/locale.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Locale = exports.make = void 0;
exports.make = (basicLocale, locales) => new Locale(basicLocale, locales);
class Locale {
    constructor(basicLocale, locales) {
        var _a;
        this.basicLocale = basicLocale;
        this.locales = locales;
        //type KeyType = keyof typeof localeEn;
        this.localeTableKey = JSON.parse(process.env.VSCODE_NLS_CONFIG).locale;
        this.localeTable = Object.assign(JSON.parse(JSON.stringify(this.basicLocale)), ((_a = this.locales[this.localeTableKey]) !== null && _a !== void 0 ? _a : {}));
        this.isTypealbeLocale = [
            "zh-CN",
            "zh-TW",
            "ja",
            "ko",
        ].indexOf(this.localeTableKey) < 0;
        this.string = (key) => this.localeTable[key] || key;
        this.map = (key) => this.string(key);
        this.typeableMap = (key) => this.isTypealbeLocale ?
            this.string(key) :
            `${this.string(key)} ( ${this.basicLocale[key]} )`;
    }
}
exports.Locale = Locale;
